### PR TITLE
Use logical line numbers in the gutter

### PR DESCRIPTION
**Pull request by cmyr**
This is frontend support for https://github.com/xi-editor/xi-editor/pull/1056; I'll update the submodule when that is merged.

## Review Checklist

- [ ] I have responded to reviews and made changes where appropriate.
- [x] I have tested the code
- [x] I have updated comments / documentation related to the changes I made.
- [x] I have rebased my PR branch onto xi-mac/master.

### CODE REVIEW

File: Sources/XiEditor/LineCache.swift [hunk 1]
```diff
@@ -170,7 +174,20 @@ fileprivate class LineCacheState<T>: UnfairLock {
                     }
                     let startIx = oldIx - nInvalidBefore
                     if op_type == "copy" {
-                        newLines.append(contentsOf: lines[startIx ..< startIx + nCopy])
+                        var lineNumber = op["ln"] as! UInt
+                        let toCopy = lines[startIx ..< startIx + nCopy]
+                        // the line number in the update is the logical line number of the first VISUAL
+                        // line to copy. If this line is not itself logical, increment lineNumber for the next line.
+                        if toCopy.first??.number == nil {
```
Comment (jansol) [inline Sources/XiEditor/LineCache.swift]: Am I missing something about Swift syntax or should this be `first?.number`?

Comment (jansol) [reply Sources/XiEditor/LineCache.swift]: Hmm, or is this actually `toCopy.first ?? (self).number`? In that case, please reformat.

**Comment by jansol** [comment]
Hm, upon further inspection it seems that the current line number in the gutter is only highlighted if the cursor is on the first visual part of that logical line. Not sure how easy/hard that would be to fix.

File: Sources/XiEditor/LineCache.swift (hunk 1 above)
Comment (cmyr) [reply Sources/XiEditor/LineCache.swift]: no, this is a double optional; the items in this list are themselves option types, and the `first` accessor returns an option (the list may be empty). I did a double take as well.

**Comment by cmyr** [comment]
That's a good point about the highlighting the line number; let's open an issue for that when this gets in.

File: Sources/XiEditor/LineCache.swift (hunk 1 above)
Comment (jansol) [reply Sources/XiEditor/LineCache.swift]: Care to add a comment :D I'm sure I'll be confused again next week if I get back to this code.

**Pull request merged**